codegenerator/cli/templates/static/codegen/src/eventFetching/ChainManager.res

```diff
@@ -257,25 +257,63 @@ let popBatchItem = (
     switch maybeArbItem {
     //If there is item on the arbitray events queue, and it is earlier than
     //than the earlist event, take the item off from there
-    | Some({val: itemWithPopFn, isInReorgThreshold})
-      if Item(itemWithPopFn)->getQueueItemComparitor(~chain=itemWithPopFn.item.chain) <
-        earliestEvent->getQueueItemComparitor(~chain) => {
+    | Some({val: arbVal, isInReorgThreshold})
+      if if arbVal.item.chain === chain {
+        // This is for the case when FetchState has a NoItem
+        // if we use a multichain comparison, the arbVal will be ignored because of lower timestamp
+        // prevent it by a special case for events with the same chain
+        Item(arbVal)->FetchState.qItemLt(earliestEvent)
+      } else {
+        arbVal.item->getComparitorFromItem < earliestEvent->getQueueItemComparitor(~chain)
+      } => {
         isInReorgThreshold,
-        val: Some(itemWithPopFn),
+        val: Some(arbVal),
       }
     | _ =>
       switch earliestEvent {
-      | NoItem(_) => {
-          isInReorgThreshold,
-          val: None,
+      | NoItem(_) =>
+        // Case when we don't have items in fetch state for any chain
+        // but there's an arbitary event on one of the chains
+        if (
+          isUnorderedMultichainMode &&
+          // This check is needed to prevent comparing items for the second time
+          maybeArbItem->Option.isNone
+        ) {
+          switch arbitraryEventQueue->getFirstArbitraryEventsItem(~fetchStatesMap) {
+          | Some({val: anotherChainArbVal, isInReorgThreshold})
+            if {
+              let anotherChain = anotherChainArbVal.item.chain
+              if anotherChain === chain {
```

Comment:
```suggestion
              let anotherChain = anotherChainArbVal.item.chain
              // This is unreachable because of the maybeArbItem->Option.isNone check above
              // Keep it just in case the logic is changed
              if anotherChain === chain {
```